lib/auth_app/accounts.ex

```diff
@@ -234,6 +261,60 @@ defmodule AuthApp.Accounts do
     Repo.one(query)
   end
 
+  @doc """
+  Gets the user with the given magic link token.
+  """
+  def get_user_by_magic_link_token(token) do
+    with {:ok, query} <- UserToken.verify_magic_link_token_query(token),
+         {user, _token} <- Repo.one(query) do
+      user
+    else
+      _ -> nil
+    end
+  end
+
+  @doc """
+  Gets the user with the given signed token.
+  """
+  def magic_link_sign_in(token) do
+    {:ok, query} = UserToken.verify_magic_link_token_query(token)
+
+    case Repo.one(query) do
+      # TODO: check if we really want to mix confirmation and login
+      {%User{email: current_email, confirmed_at: nil} = user, %UserToken{sent_to: current_email}} ->
+        {:ok, %{user: user, tokens_to_expire: tokens_to_expire}} =
+          Repo.transaction(confirm_user_multi(user))
+
+        {:ok, user, tokens_to_expire}
+
+      {user, token} ->
+        Repo.delete!(token)
+        {:ok, user, []}
+
+      nil ->
+        {:error, :not_found}
+    end
+  end
+
+  @doc """
+  Checks whether the user is in sudo mode.
+
+  The user is in sudo mode when the last authentication was done no further
+  than 15 minutes ago.
+  """
+  def sudo_mode?(%User{authenticated_at: ts}) when is_struct(ts, DateTime) do
+    DateTime.after?(ts, DateTime.utc_now() |> DateTime.add(-15, :minute))
```

Comment:
we could probably use `System.monotonic_time(:second)` instead, which would be less expensive?

Reply:
Yes. We can also use duration if we want: `System.monotonic_time(:microsecond) + to_timeout(minute: 15)`

Reply:
actually we cannot, as the we use the token's inserted_at for authenticated_at. And there's no way to translate a DateTime into monotonic time.